### PR TITLE
gh-118911: Trailing whitespace in a block shouldn't prevent the user from terminating the code block

### PR DESCRIPTION
**Pull request by aelsayed95**
Issue described in https://github.com/python/cpython/issues/118911

Fix includes:
1. If the user hits enter twice, they are able to exit out of the block, even if there's trailing whitespace. 
2. Trim trailing whitespace in the `HistoricalReader` to match IPython's behavior  

* Issue: gh-118911

### CODE REVIEW

**Comment by aelsayed95** [comment]
One issue discovered while testing this (exists on main too):

Within a code block, if the user hits enter twice to add new lines, the first line matches the previous line's indentation while the next one doesn't.
This doesn't match IPython's behaviour.

**Comment by aelsayed95** [comment]
Added https://github.com/python/cpython/issues/119356

File: Misc/NEWS.d/next/Library/2024-05-21-20-13-23.gh-issue-118911.iG8nMq.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+To match IPython's behaviour, updated `maybe-accept`'s logic so that if the user hits enter twice, they are able to terminate the block even if there's trailing whitespace.
+Also, trimming trailing whitespace in HistoricalReader so that when the user hits arrow up, the cursor is on the last functional line. This matches IPython's behavior.
```
Comment (savannahostrowski) [inline Misc/NEWS.d/next/Library/2024-05-21-20-13-23.gh-issue-118911.iG8nMq.rst]: It looks like you might have a redundant "This matches IPython's behavior." at the end of this news entry (as you already mention this at the beginning of the blurb).

File: Lib/_pyrepl/readline.py [hunk 2]
```diff
@@ -219,16 +219,29 @@ def do(self) -> None:
...
-            if not self.reader.paste_mode and indent:
-                for i in range(prevlinestart, prevlinestart + indent):
-                    r.insert(r.buffer[i])
+            # if there's already a new line before the cursor then
+            # even if the cursor is followed by whitespace, we assume
+            # the user is trying to terminate the block
+            if _newline_before_pos() and text[r.pos:].isspace():
+                self.finish = True
+            else:
+                #
```
Comment (savannahostrowski) [inline Lib/_pyrepl/readline.py]: Just out of curiosity, I see some other places in the file that use these empty comment lines before starting the comment block...Is this a style convention? I haven't really seen it used elsewhere.

**Review by savannahostrowski — COMMENTED** [review]
Nice! I tested this on Fedora (via Devcontainer) and it looks like this now works as described.

File: Lib/_pyrepl/readline.py (hunk 2 above)
Comment (ambv) [reply Lib/_pyrepl/readline.py]: Looks like this was an original style of pyrepl as adopted from PyPy. I agree with you Savannah that it looks out of place. We probably want to get rid of it.

File: Misc/NEWS.d/next/Library/2024-05-21-20-13-23.gh-issue-118911.iG8nMq.rst (hunk 1 above)
Comment (ambv) [reply Misc/NEWS.d/next/Library/2024-05-21-20-13-23.gh-issue-118911.iG8nMq.rst]: Agree with Savannah, but actually there's another thing here to consider.

We can say that "This change matches IPython's behavior" to communicate precedent in the chosen design. However, we don't want to say that we're making changes "to match IPython's behavior". This is not a design goal for us. It's too deep of a rabbit hole to commit to replicating IPython ☺️

**Review by ambv — APPROVED** [review]
This looks good.

I merged new changes in `main` into your PR since there were conflicts, and reworded the blurb per my suggestion.

**Comment by ambv** [comment]
Thanks, Aya! ✨ 🍰 ✨

**Comment by ambv** [comment]
The backport depends on GH-119362. I'll deal with that tomorrow.

**Pull request merged**